### PR TITLE
Approximate Pruning while doing incremental inference

### PR DESCRIPTION
**Pull request by ProfFan**
This PR does:
  - Add `nrLeaves` in `DecisionTree`.
  - Simpler printing code (@varunagrawal)
  - Add `nrComponents` in `GaussianMixture`
  - Pruning when we incrementally update

The pruning is done by
1. Thresholding the discrete factor with probability less than rank-N
2. Replace the corresponding components with nullptrs in the GaussianMixture
3. In elimination make sure that nullptrs are dealt with

### CODE REVIEW

File: gtsam/discrete/DecisionTree.h
Comment (dellaert) [inline gtsam/discrete/DecisionTree.h:44]: @varunagrawal this change is here only because your PR has not removed the superfluous decisiontree changes yet :-) Could you do that first before I (or you!) go over this PR?

File: gtsam/discrete/DiscreteFactorGraph.h
Comment (dellaert) [inline gtsam/discrete/DiscreteFactorGraph.h:176]: undo accidental edit

File: gtsam/hybrid/IncrementalHybrid.cpp
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.cpp:55]: format w Google style and 80 char limit

**Review by dellaert — APPROVED** [review]
Approved with some nots, but please wait until base branch is merged...

File: gtsam/discrete/DecisionTree.h
Comment (varunagrawal) [reply gtsam/discrete/DecisionTree.h:44]: I definitely resolved those. :/

**Comment by dellaert** [comment]
Conflicts and failing CI

**Comment by ProfFan** [comment]
https://godbolt.org/z/4n4ETqKqs is the proper way to do equals

**Pull request merged**
